### ISSUE TITLE
Using Postgresql with Prepared Statements turned off, #exec_query discards bind parameters

### ISSUE THREAD

**Issue by joecannatti** [rails/rails#24893]
##### Create a new rails app using Psql

```
$ rails new -d postgresql reproduce-app
$ cd reproduce-app/
```
##### Configure the database to turn off prepared_statements

```
default: &default
  adapter: postgresql
  encoding: unicode
  prepared_statements: false
  username: your_user
  password: your_password
```
#### Create a very simple model to test against

```
rails g model User name:string
rake db:create db:migrate db:test:prepare
```
#### Use `#exec_query` with bind params to query the table

```
 $ rails c
 Running via Spring preloader in process 12847
 Loading development environment (Rails 4.2.5.2)
 2.2.2 :001 > User.connection.exec_query("SELECT * from users where name = $1", "users", [[nil, 'aoeu']])
 ActiveRecord::StatementInvalid: PG::ProtocolViolation: ERROR:  bind message supplies 0 parameters, but prepared statement "" requires 1
```
### Expected behavior

The query to be run successfully
### Actual behavior

```
 ActiveRecord::StatementInvalid: PG::ProtocolViolation: ERROR:  bind message supplies 0 parameters, but prepared statement "" requires 1
```
### System configuration

**Rails version**: Rails 4.2.5.2

**Ruby version**: 2.3.0
### My attempt at resolving the issue

This comes down to the method `ActiveRecord::ConnectionAdapters::PostgreSQLAdapter#exec_no_cache`, which throws out the `binds` array and passes `[]` in it's place

```
# File activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb, line 591
        def exec_no_cache(sql, name, binds)
          log(sql, name, binds) { @connection.async_exec(sql, []) }
        end
```

I tried monkey patching that like so ->

```
# File activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb, line 591
        def exec_no_cache(sql, name, binds)
          type_casted_binds = binds.map { |col, val|
            [col, type_cast(val, col)]
          }

          log(sql, name, binds) { @connection.async_exec(sql, type_casted_binds.map { |_, val| val }) }
        end
```

**Which did fix the error.**

```
2.3.0 :010 > User.connection.exec_query("SELECT * from users where name = $1", "users", [[nil, 'aoeu']])
  users (11.7ms)  SELECT * from users where name = $1  [[nil, "aoeu"]]
#SUCCESS!!!
```

But it breaks `#create`

```
 User.create(name: "aoeu")
   (0.2ms)  BEGIN
  SQL (0.6ms)  INSERT INTO "users" ("name", "created_at", "updated_at") VALUES ('aoeu', '2016-05-06 15:28:23.694039', '2016-05-06 15:28:23.694039') RETURNING "id"  [["name", "aoeu"], ["created_at", Fri, 06 May 2016 15:28:23 UTC +00:00], ["updated_at", Fri, 06 May 2016 15:28:23 UTC +00:00]]
   (0.1ms)  ROLLBACK
ActiveRecord::StatementInvalid: PG::IndeterminateDatatype: ERROR:  could not determine data type of parameter $1
: INSERT INTO "users" ("name", "created_at", "updated_at") VALUES ('aoeu', '2016-05-06 15:28:23.694039', '2016-05-06 15:28:23.694039') RETURNING "id"
        from (irb):6:in `async_exec'
        from (irb):6:in `block in exec_no_cache'
        from /home/joecannatti/.rvm/gems/ruby-2.3.0/gems/activerecord-4.2.5.2/lib/active_record/connection_adapters/abstract_adapter.rb:472:in `block in log'
        from /home/joecannatti/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.5.2/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
        from /home/joecannatti/.rvm/gems/ruby-2.3.0/gems/activerecord-4.2.5.2/lib/active_record/connection_adapters/abstract_adapter.rb:466:in `log'
```

My questions are
- Should I be able to use `User.connection.exec_query("SELECT * from users where name = $1", "users", [[nil, 'aoeu']])` with prepared statements turned off?
  - If so, can you provide me some of the nuance/details of the internals to help me make the fix?
  - If not, should the error message be improved?

**Comment by vipulnsward** [comment]
cc @sgrif

**Comment by joecannatti** [comment]
Just want to make sure that this didn't get lost because it was opened during Railsconf.

**Comment by sgrif** [comment]
It wasn't lost

**Comment by sgrif** [comment]
<img width="253" alt="screen shot 2016-05-12 at 12 27 10 am" src="https://cloud.githubusercontent.com/assets/1529387/15204151/61a37db4-17d8-11e6-8367-39fa56c61d14.png">

Your patience is appreciated though. There's a backlog to work through, and I'm not officially back until Monday. :)

**Comment by sgrif** [comment]
Can you see if this issue is reproducible in 5.0.0.rc1? I believe this has already been fixed

**Comment by dnasseri** [comment]
@sgrif I ran into the same bug, and reported it here: https://github.com/rails/rails/issues/25995.

I think the issue is still present in rc1. Here is a test to reproduce:

```
require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'
  gem 'activerecord', '5.0.0.rc1'
  gem 'pg'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

ActiveRecord::Base.establish_connection(
  adapter: 'postgresql',
  port: 5432,
  prepared_statements: false
)

ActiveRecord::Base.logger = Logger.new(STDOUT)
ActiveRecord::Schema.define do
  create_table :people, temporary: true do |t|
    t.string :name
  end
end

class Person < ActiveRecord::Base; end

class BugTest < Minitest::Test

  def test_error
    person = Person.new
    person.name = 'FooBar'
    person.save!
    rows = ActiveRecord::Base.connection.exec_query(
      'SELECT name FROM people WHERE id = $1',
      'sql',
      [person.id]
    )
    assert rows.length = 1
  end

end
```

**Comment by dnasseri** [comment]
@sgrif I am also curios about whether this is expected to work when prepared statements are disabled. If not, I would agree with @joecannatti that the error messages could be improved. 

Here is a test case on AR 3.2.22 which does not error:

```
require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'
  gem 'activerecord', '3.2.22'
  gem 'minitest'
  gem 'pg'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

ActiveRecord::Base.establish_connection(
  adapter: 'postgresql',
  database: 'postgres',
  prepared_statements: false
)

#ActiveRecord::Base.logger = Logger.new(STDOUT)
ActiveRecord::Schema.define do
  create_table :people, temporary: true do |t|
    t.string :name
  end
end

class Person < ActiveRecord::Base; end

class BugTest < Minitest::Test

  def test_no_error
    person = Person.new
    person.name = 'FooBar'
    person.save!
    rows = ActiveRecord::Base.connection.exec_query(
      'SELECT name FROM people WHERE id = $1',
      'sql',
      [person.id]
    ).rows
  end

end
```

As a workaround I have been using find_by_sql on an arbitrary model in rails 4, but it seems unnecessary to generate all those model instances. 

Thanks so much for the help!

**Comment by sam2000** [comment]
I am having the same issue in rails 4 as well.  I am using string to get stuff done but I guess it is bad for performance as it is a frequently used query.  I do you this gets fixed soon.

**Comment by sgrif** [comment]
So passing binds here is going to be relying on internal APIs no matter what, as their representation isn't considered public API. In Rails 5, you'll need to pass `prepare: true` to `exec_query`, and your bind parameters will need to be objects which respond to `value_for_database`. Beyond that, there doesn't seem to be any bug here.

**Comment by xaviershay** [comment]
This may still technically be "not a bug because private API", but the suggested approach by @sgrif  does not appear to work on 5.2.3. Amended repro above to set `prepare: true` and use appropriate types for binds.

```
require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'
  gem 'activerecord', '5.2.3'
  gem 'pg'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

ActiveRecord::Base.establish_connection(
  adapter: 'postgresql',
  port: 5432,
  prepared_statements: false
)

ActiveRecord::Base.logger = Logger.new(STDOUT)
ActiveRecord::Schema.define do
  create_table :people, temporary: true do |t|
    t.string :name
  end
end

class Person < ActiveRecord::Base; end

class BugTest < Minitest::Test

  def test_error
    binds = [
      ActiveRecord::Relation::QueryAttribute.new("id", 1, ActiveRecord::Type::Integer.new)
    ]

    binds.each do |bind|
      assert bind.respond_to?(:value_for_database)
    end

    person = Person.new
    person.name = 'FooBar'
    person.save!
    rows = ActiveRecord::Base.connection.exec_query(
      'SELECT name FROM people WHERE id = $1',
      'sql',
      binds,
      prepare: true
    )
  end

end
```

I am still searching for a workaround.

**Comment by xaviershay** [comment]
In `postgres_adapter.rb` for 5.2.3, the `prepare` parameter is ignored if prepared statements are disabled globally. This feels unexpected?

```
        def execute_and_clear(sql, name, binds, prepare: false)
          if without_prepared_statement?(binds)
            result = exec_no_cache(sql, name, [])
          elsif !prepare
            result = exec_no_cache(sql, name, binds)
          else
            result = exec_cache(sql, name, binds)
          end
          ret = yield result
          result.clear
          ret
        end
```

I _think_ the behaviour I want is but haven't validated much yet:

```
        def execute_and_clear(sql, name, binds, prepare: false)
          if without_prepared_statement?(binds)
            result = exec_no_cache(sql, name, [])
          elsif !prepare
            result = exec_no_cache(sql, name, binds)
          else
            result = exec_cache(sql, name, binds)
          end
          ret = yield result
          result.clear
          ret
        end
```

**Comment by ashkulz** [comment]
This was fixed via #41300 and back-ported to 6.1 via f5a87fede3bc77ef938bdb015f2866b3cd8db662.
